go/service/usersearch.go

```diff
@@ -188,33 +197,159 @@ func contactSearch(mctx libkb.MetaContext, arg keybase1.UserSearchArg) (res []ke
+		}
+
+		for _, v := range lookupRes {
+			// Found a resolution
+			if v.Error != "" || v.UID.IsNil() {
+				continue
+			}
+
+			assertionValue := queryString
+			if v.Coerced != "" {
+				assertionValue = v.Coerced
+			}
+			imptofu := &keybase1.ImpTofuSearchResult{
+				CoercedQuery: assertionValue,
+				Resolved:     true,
+				Uid:          v.UID,
+				Assertion:    imptofuQueryToAssertion(imptofuType, assertionValue),
+				Username:     v.UID.String(),
+			}
+			if following != nil {
+				imptofu.Following = following[v.UID]
+			}
+			if usernames != nil {
+				if uname, found := usernames[v.UID]; found {
+					imptofu.Username = uname.Username
+					imptofu.FullName = uname.Fullname
+				}
+			}
+			res = &keybase1.APIUserSearchResult{
+				Score:   1.0,
+				Imptofu: imptofu,
+			}
+			return res, nil // return here - we only want one result
+		}
+	}
+
+	// Not resolved - add SBS result.
+	imptofu := &keybase1.ImpTofuSearchResult{
+		CoercedQuery: queryString,
+		Resolved:     false,
+		Assertion:    imptofuQueryToAssertion(imptofuType, queryString),
+	}
+	res = &keybase1.APIUserSearchResult{
+		Score:   1.0,
+		Imptofu: imptofu,
+	}
+	return res, nil
+}
+
 func (h *UserSearchHandler) UserSearch(ctx context.Context, arg keybase1.UserSearchArg) (res []keybase1.APIUserSearchResult, err error) {
 	mctx := libkb.NewMetaContext(ctx, h.G()).WithLogTag("USEARCH")
 	defer mctx.TraceTimed(fmt.Sprintf("UserSearch#UserSearch(s=%q, q=%q)", arg.Service, arg.Query),
 		func() error { return err })()
 
 	if arg.Query == "" {
-		return res, nil
+		return nil, nil
+	}
+
+	if !h.disableServiceSearch {
+		res, err = doSearchRequest(mctx, arg)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	res, err = doSearchRequest(mctx, arg)
 	if arg.IncludeContacts {
 		contactsRes, err := contactSearch(mctx, arg)
 		if err != nil {
-			return nil, err
+			mctx.Error("Failed to search synced contacts: %s", err)
+		} else if len(contactsRes) > 0 {
+			res = append(contactsRes, res...)
 		}
-		if len(contactsRes) > 0 {
-			var res2 []keybase1.APIUserSearchResult
-			res2 = append(res2, contactsRes...)
-			res2 = append(res2, res...)
-			res = res2
-
-			maxRes := arg.MaxResults
-			if maxRes > 0 && len(res) > maxRes {
-				res = res[:maxRes]
+	}
+
+	if arg.ImpTofuQuery != nil {
+		imptofuRes, err := imptofuSearch(mctx, h.contactsProvider, *arg.ImpTofuQuery)
+		if err != nil {
+			mctx.Error("Failed to do phone number / email search: %s", err)
+		} else if imptofuRes != nil {
+			// Check if we have found assertion of this result already in our
+			// contacts.
+			var found bool
+			for _, v := range res {
+				if v.Contact != nil && v.Contact.Assertion == imptofuRes.Imptofu.Assertion {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				// Prepend *imptofuRes
+				res = append([]keybase1.APIUserSearchResult{*imptofuRes}, res...)
 			}
 		}
 	}
 
+	maxRes := arg.MaxResults
+	if maxRes > 0 && len(res) > maxRes {
```

Comment:
what is this value usually going to be? can we save work by not doing a bunch of service queries and so on before if we're going to throw them away later?

Reply:
you're right ill just redo this function - reorder the calls to allow us skip some of them if we are over the limit of search result anyways.

i don't know what i was thinking when i've done the reverse order calls with prepending to the list, maybe there was a reason, i need to think

Reply:
This is postponed till later - also we are sorting this list differently now so we most likely always have to ask for full result set from API.

Reply:
Ok, can you put a comment with ticket number?